enterprise/server/quota/quota_manager.go

```diff
@@ -105,66 +104,74 @@ func fetchConfigFromDB(env environment.Env, namespace string) (map[string]*names
 	ctx := env.GetServerContext()
 	config := make(map[string]*namespaceConfig)
-	// TODO(zoey): use JOIN for this
-	err := env.GetDBHandle().Transaction(ctx, func(tx interfaces.DB) error {
-		q := query_builder.NewQuery(`SELECT * FROM "QuotaBuckets"`)
-		if namespace != "" {
-			q = q.AddWhereClause(`namespace = ?`, namespace)
-		}
-		queryStr, args := q.Build()
-		rq := tx.NewQuery(ctx, "quota_manager_get_quota_buckets").Raw(queryStr, args...)
-		err := db.ScanEach(rq, func(ctx context.Context, tb *tables.QuotaBucket) error {
-			if err := validateBucket(bucketToProto(tb)); err != nil {
-				return status.InternalErrorf("invalid bucket: %v", tb)
-			}
-			ns := config[tb.Namespace]
-			if ns == nil {
+	var rq interfaces.DBRawQuery
+	if namespace == "" {
+		rq = env.GetDBHandle().NewQuery(ctx, "quota_manager_fetch_all_configs_from_db").Raw(`
+			SELECT qb.*, qg.*
+			FROM "QuotaBuckets" AS qb
+				LEFT JOIN "QuotaGroups" AS qg
+					ON qb.namespace = qg.namespace
+					AND qb.name = qg.bucket_name
+			ORDER BY
+				qb.namespace,
+				qb.name ASC,
+				qg.quota_key ASC`,
+		)
+	} else {
+		rq = env.GetDBHandle().NewQuery(ctx, "quota_manager_fetch_config_from_db").Raw(`
+			SELECT qb.*, qg.*
+			FROM "QuotaBuckets" AS qb
+				LEFT JOIN "QuotaGroups" AS qg
+					ON qb.namespace = qg.namespace
+					AND qb.name = qg.bucket_name
+			WHERE qb.namespace = ?
+			ORDER BY
+				qb.name ASC,
+				qg.quota_key ASC`,
+			namespace,
+		)
+	}
+	err := db.ScanEach(
+		rq,
+		func(ctx context.Context, qbg *struct {
+			tables.QuotaBucket
+			tables.QuotaGroup
+		}) error {
+			qb := &qbg.QuotaBucket
+			ns, ok := config[qb.Namespace]
+			if !ok {
 				ns = &namespaceConfig{
-					name:            tb.Namespace,
+					name:            qb.Namespace,
 					assignedBuckets: make(map[string]*assignedBucket),
 				}
-				config[tb.Namespace] = ns
+				config[qb.Namespace] = ns
 			}
-			ns.assignedBuckets[tb.Name] = &assignedBucket{
-				bucket: tb,
+			bucket, ok := ns.assignedBuckets[qb.Name]
+			if !ok {
+				if err := validateBucket(bucketToProto(qb)); err != nil {
+					return status.InternalErrorf("invalid bucket: %v", qbg.QuotaBucket)
+				}
+				bucket = &assignedBucket{
+					bucket: &qbg.QuotaBucket,
+				}
+				ns.assignedBuckets[qb.Name] = bucket
 			}
-			return nil
-		})
-		if err != nil {
-			return status.InternalErrorf("failed to read table QuotaBuckets: %s", err)
-		}
 
-		groupQuery := query_builder.NewQuery(`SELECT * FROM "QuotaGroups"`)
-		if namespace != "" {
-			groupQuery = groupQuery.AddWhereClause(`namespace = ?`, namespace)
-		}
-		groupQueryStr, groupArgs := groupQuery.Build()
-		rq = tx.NewQuery(ctx, "quota_manager_get_quota_groups").Raw(groupQueryStr, groupArgs...)
-		err = db.ScanEach(rq, func(ctx context.Context, tg *tables.QuotaGroup) error {
-			ns := config[tg.Namespace]
-			if ns == nil {
-				alert.UnexpectedEvent("invalid_quota_config", "namespace %q doesn't exist", tg.Namespace)
+			qg := &qbg.QuotaGroup
+			if qg.Namespace == "" {
```

Comment:
sorry, this is mostly just to make sure i understand what's going on here--is this even possible? what does it mean?  it seems like this check isn't present in the original code, but it also seems harmless.

Reply:
I've corrected it to use a pointer for the quota group and now check for nil quota group (which I only recently learned was a possibility); it meant there was no quota group for this quota bucket, so we don't want to add an empty quota key to the quota keys. Now we check for nil and do the same thing. It's different than the original code because previously no row would be returned if there were no corresponding quota group, but the left join means that we will still get a row for the bucket with an empty group if no quota group exists, so we have to explicitly skip.